urlc8dbc7ce6e_dlubarov_guava_tgz-pJ8-MainJ8/src/c/TypeDef.java

```diff
@@ -0,0 +1,441 @@
+            ParameterizedType thisType;
+            try {
+                thisType = argTypes[0].asSupertype(desc, ctx);
+            } catch (IllegalArgumentException e) {
+                break instanceMethodSearch;
+            }
+
+            Type[] otherArgTypes = new Type[argTypes.length - 1];
+            System.arraycopy(argTypes, 1, otherArgTypes, 0, otherArgTypes.length);
+            try {
+                options.add(getInstanceMethod(name, thisType.genericArgs, methodGenerics, otherArgTypes, ctx));
+            } catch (NoSuchElementException e) {}
+        }
+
+        methodSearch:
+        for (MethodDef method : allMethodDefs) {
+            if (!name.equals(method.name))
+                continue;
+
+            // TODO: allow static invocation of instance methods
+            if (!method.isStatic)
+                continue;
+
+            if (methodGenerics.length != method.genericInfos.length)
+                continue;
+
+            Type[] expectedTypes = method.paramTypes;
+            if (argTypes.length != expectedTypes.length)
+                continue;
+
+            for (int i = 0; i < expectedTypes.length; ++i) {
+                Type expectedType = expectedTypes[i];
+                Type expectedTypeWithGens = expectedType.withGenericArgs(null, methodGenerics);
+                // TODO: enforce generic bounds
+                if (!argTypes[i].isSubtype(expectedTypeWithGens, null, ctx.method))
+                    continue methodSearch;
+            }
+
+            options.add(method);
+        }
+
+        // There should be exactly one matching method, which we return.
+        if (options.isEmpty())
+            throw new NoSuchElementException(String.format(
+                    "No matching static method found for %s.%s with argument types %s.",
+                    desc, name, Arrays.toString(argTypes)));
+        if (options.size() > 1)
+            throw new NiftyException("Ambiguous static method call '%s'.", name);
+        return options.iterator().next();
+    }
+
+    public FieldDef getInstanceField(String name) {
+        Set<FieldDef> options = new HashSet<FieldDef>();
+
+        // Search my own fields.
+        for (FieldDef field : instanceFieldDefs)
+            if (field.name.equals(name))
+                options.add(field);
+
+        // Search my parents' fields.
+        for (ParameterizedType parent : parents)
+            try {
+                options.add(Project.singleton.resolve(parent.rawType).getInstanceField(name));
+            } catch (NoSuchElementException e) {}
+
+        // There should be exactly one matching field.
+        if (options.isEmpty())
+            throw new NoSuchElementException(desc + " has no instance field named " + name);
+        if (options.size() > 1)
+            throw new NiftyException("%s has multiple fields named %s", desc, name);
+        return options.iterator().next();
+    }
+
+    public MethodDef getInstanceMethod(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 187-194 into a separate method.